src/charts/chart.js

```diff
@@ -281,110 +271,28 @@ export default function Chart (_container) {
       .style("width", `${config.markPanelWidth}px`)
       .style("height", `${config.chartHeight}px`)
       .attr("fill", "transparent")
-
     return this
   }
 
   function buildChart () {
-    scales = components.scale
+    config = transformConfig(cache.config)
+    data = transformData(cache.data)
+    scales = scale
       .setConfig(config)
-      .setData(dataObject)
+      .setData(data)
       .getScales()
 
-    components.clipPath
-      .setConfig(config)
-      .render()
-
-    components.axis
-      .setConfig(config)
-      .setScales(scales)
-      .render()
-
-    components.bar
-      .setConfig(config)
-      .setScales(scales)
-      .setData(dataObject)
-      .render()
-
-    components.line
-      .setConfig(config)
-      .setScales(scales)
-      .setData(dataObject)
-      .render()
+    update()
 
-    components.tooltip
-      .setConfig(config)
-      .setScales(scales)
-      .bindEvents(dispatcher)
-
-    components.legend
-      .setConfig(config)
-      .setScales(scales)
-      .setData(dataObject)
-
-    components.brush
-      .setConfig(config)
-      .setScales(scales)
-      .setData(dataObject)
-      .render()
-
-    components.hover
-      .setConfig(config)
-      .setScales(scales)
-      .setData(dataObject)
-      .bindEvents(dispatcher)
-
-    components.binning
-      .setConfig(config)
-      .render()
-
-    components.domainEditor
-      .setConfig(config)
-      .setScales(scales)
-      .render()
-
-    components.brushRangeEditor
-      .setConfig(config)
-      .setScales(scales)
-      .render()
-
-    components.label
-      .setConfig(config)
-      .render()
-
-    triggerIntroAnimation()
-    return this
-  }
-
-  function setData (_data) {
-    dataObject.data = cloneData(_data[keys.SERIES])
-    const cleanedData = dataManager.cleanData(_data, config.keyType, config.sortBy, config.fillData)
-    Object.assign(dataObject, cleanedData)
-
-    const autoConfig = autoConfigure(inputConfig, cache, dataObject)
-    config = Object.assign({}, inputConfig, autoConfig)
-
-    render()
+    componentRegistry.render({
```

Comment:
This does seem cleaner or like a good refactor. 

I'm wondering why it's necessary to pass an `options` object to the render method at all, won't these properties always be needed when calling render? Or is the intention to allow for disabling some of them in the future?

Reply:
The goal of this componentRegistry is to replace a long series of 
```
component
    .setConfig(config)
    .setScales(scales)
    .setData(dataObject)
    .render()
```
With a single call equivalent to 
```
allComponent
    .setConfig(config)
    .setScales(scales)
    .setData(dataObject)
    .render()
```
The function signature is just different
```
componentRegistry.render({
      config,
      scales,
      data,
      dispatcher
    })
```
Maybe it would be better to keep it closer to the component's signature?
```
componentRegistry
    .setConfig(config)
    .setScales(scales)
    .setData(data)
    .bindEvents(dispatcher)
    .render()
```

Reply:
hmmm, yeah the latter makes it a little more clear what is happening with componentRegistry. Wrapping each of those objects in a single `render` obfuscates things I think. But maybe not.